bazel/python_rules.bzl

```diff
@@ -68,73 +102,89 @@ def _generate_py_impl(context):
     if out_dir.import_path:
         imports.append("%s/%s/%s" % (context.workspace_name, context.label.package, out_dir.import_path))
 
+    py_info = PyInfo(transitive_sources = depset(direct = out_files), imports = depset(direct = imports))
+    return PyProtoInfo(
+        py_info = _merge_pyinfos(
+            [
+                py_info,
+                context.attr._protobuf_library[PyInfo],
+            ] + [dep[PyProtoInfo].py_info for dep in context.rule.attr.deps],
+        ),
+        generated_py_srcs = generated_py_srcs,
+    )
+
+_gen_py_aspect = aspect(
+    implementation = _gen_py_aspect_impl,
+    attr_aspects = ["deps"],
+    fragments = ["py"],
+    attrs = {
+        "_protoc": attr.label(
+            default = Label("//external:protocol_compiler"),
+            providers = ["files_to_run"],
+            executable = True,
+            cfg = "host",
+        ),
+        "_protobuf_library": attr.label(
+            default = Label("@com_google_protobuf//:protobuf_python"),
+            providers = [PyInfo],
+        ),
+    },
+)
+
+def _generate_py_impl(context):
+    if (len(context.attr.deps) != 1):
+        fail("Can only compile a single proto at a time.")
+
+    py_sources = []
+
+    # If the proto_library this rule *directly* depends on is in another
+    # package, then we generate .py files to import them in this package. This
+    # behavior is needed to allow rearranging of import paths to make Bazel
+    # outputs align with native python workflows.
+    if context.label.package != context.attr.deps[0].label.package:
+        for py_src in context.attr.deps[0][PyProtoInfo].generated_py_srcs:
+            reimport_py_file = context.actions.declare_file(py_src.basename)
+            py_sources.append(reimport_py_file)
+            import_line = "from %s import *" % py_src.short_path.replace("/", ".")[:-3]
```

Comment:
It's worth noting that this approach could theoretically be defeated if `protoc` started defining `__all__` or started defining symbols with a `__` prefix. However, if such symbols did exist, they would almost certainly be considered private API, so it's probably not worth worrying about.